applications/nrf5340_audio/src/audio_module/src/amod_api.h

```diff
@@ -0,0 +1,496 @@
+#include <zephyr/kernel.h>
+
+#include "data_fifo.h"
+#include "ablk_api.h"
+
+/**
+ * @brief Module type.
+ */
+enum amod_type {
+	/* The module type is undefined */
+	AMOD_TYPE_UNDEFINED = 0,
+
+	/* This is an input processing module */
+	AMOD_TYPE_INPUT,
+
+	/* This is an output processing module */
+	AMOD_TYPE_OUTPUT,
+
+	/* This is an input/output processing module */
+	AMOD_TYPE_IN_OUT
+};
+
+/**
+ * @brief Module state.
+ */
+enum amod_state {
+	/* Module state undefined */
+	AMOD_STATE_UNDEFINED = 0,
+
+	/* Module is in the configured state */
+	AMOD_STATE_CONFIGURED,
+
+	/* Module is in the running state */
+	AMOD_STATE_RUNNING,
+
+	/* Module is in the stopped state */
+	AMOD_STATE_STOPPED
+};
+
+/**
+ * @brief Module's private handle.
+ */
+struct handle;
+
+/**
+ * @brief Private context for the module.
+ */
+struct amod_context;
+
+/**
+ * @brief Module's private configuration.
+ */
+struct amod_configuration;
+
+/**
+ * @brief Private pointer to a module's functions.
+ */
+struct amod_functions {
+	/**
+	 * @brief Function for opening a module.
+	 *
+	 * @param handle         The handle to the module instance
+	 * @param configuration  Pointer to the module's current configuration
+	 *
+	 * @return 0 if successful, error otherwise
+	 */
+	int (*open)(struct handle *handle, struct amod_configuration *configuration);
+
+	/**
+	 * @brief Function to close an open module.
+	 *
+	 * @param handle  The handle to the module instance
+	 *
+	 * @return 0 if successful, error otherwise
+	 */
+	int (*close)(struct handle *handle);
+
+	/**
+	 * @brief Function to configure a module.
+	 *
+	 * @param handle         The handle to the module instance
+	 * @param configuration  Pointer to the module's configuration to set
+	 *
+	 * @return 0 if successful, error otherwise
+	 */
+	int (*configuration_set)(struct handle *handle, struct amod_configuration *configuration);
+
+	/**
+	 * @brief Function to get the configuration of a module.
+	 *
+	 * @param handle         The handle to the module instance
+	 * @param configuration  Pointer to the module's current configuration
+	 *
+	 * @return 0 if successful, error otherwise
+	 */
+	int (*configuration_get)(struct handle *handle, struct amod_configuration *configuration);
+
+	/**
+	 * @brief Start a module.
+	 *
+	 * @param handle  The handle for the module to start
+	 *
+	 * @return 0 if successful, error otherwise
+	 */
+	int (*start)(struct handle *handle);
+
+	/**
+	 * @brief Stop a module.
+	 *
+	 * @param handle  The handle for the module to stop
+	 *
+	 * @return 0 if successful, error otherwise
+	 */
+	int (*stop)(struct handle *handle);
+
+	/**
+	 * @brief The core data processing function for the module. Can be either an
+	 *		  input, output or processor module type.
+	 *
+	 * @param handle    The handle to the module instance
+	 * @param block_rx  Pointer to the input audio block or NULL for an input module
+	 * @param block_tx  Pointer to the output audio block or NULL for an output module
```

Comment:
I might have misunderstood, but here goes: 
Should it be?
```suggestion
	 * @param block_rx  Pointer to the input audio block or NULL for an output module
	 * @param block_tx  Pointer to the output audio block or NULL for an input module
```

Reply:
I think this is right. An input module has no receive FIFO as the data is sourced within the module (e.g. USB). Similarly the output has no output FIFO as the data leaves the module via another means (e.g. I2S).

Could I update the amod_type to be:

/**
 * @brief Module type.
 */
enum amod_type {
	/* The module type is undefined */
	AMOD_TYPE_UNDEFINED = 0,

	/* This is an input processing module.
	 * Note: An input module obtains data from peripheral (e.g. I2S) and
	 *       hence has on RX FIFO.
	 */
	AMOD_TYPE_INPUT,

	/* This is an output processing module.
	 * Note: An output module outputs data to peripheral (e.g. I2S) and
	 *       hence has on TX FIFO.
	 */
	AMOD_TYPE_OUTPUT,

	/* This is a processing module.
	 * Note: A processing module takes input and outputs from another
	 *       module or application, thus having RX and TX FIFOs.
	 */
	AMOD_TYPE_PROCESS
};

This should make the comment on NULL pointers clearer?